Fix some chainstate-init-order bugs.

TheBlueMatt opened the pull request:
This does a number of things to clean up chainstate init order,
fixing some issues as it goes:

* Order chainstate init more logically - first all of the
  blocktree-related loading, then coinsdb, then
  pcoinsTip/chainActive. Only create objects as needed.

* More clearly document exactly what is and isn't called in
  -reindex and -reindex-chainstate both with comments noting
  calls as no-ops and by adding if guards.

* Move the writing of fTxIndex to LoadBlockIndex - this fixes a
  bug introduced in d6af06d68aae985436cbc942f0d11078041d121b where
  InitBlockIndex was writing to fTxIndex which had not yet been
  checked (because LoadChainTip hadn't yet initialized the
  chainActive, which would otherwise have resulted in
  InitBlockIndex being a NOP), allowing you to modify -txindex
  without reindex, potentially corrupting your chainstate!

* Rename InitBlockIndex to LoadGenesisBlock, which is now a more
  natural name for it. Also check mapBlockIndex instead of
  chainActive, fixing a bug where we'd write the genesis block out
  on every start.

* Move LoadGenesisBlock further down in init. This is a more logical
  location for it, as it is after all of the blockindex-related
  loading and checking, but before any of the UTXO-related loading
  and checking.

* Give LoadChainTip a return value - allowing it to indicate that
  the UTXO DB ran ahead of the block DB. This just provides a nicer
  error message instead of the previous mysterious
  assert(!setBlockIndexCandidates.empty()) error.

* Calls ActivateBestChain in case we just loaded the genesis
  block in LoadChainTip, avoiding relying on the ActivateBestChain
  in ThreadImport before continuing init process.

* Move all of the VerifyDB()-related stuff into a -reindex +
  -reindex-chainstate if guard. It couldn't do anything useful
  as chainActive.Tip() would be null at this point anyway.

TheBlueMatt committed:
Fix some LoadChainTip-related init-order bugs.

* Move the writing of fTxIndex to LoadBlockIndex - this fixes a
  bug introduced in d6af06d68aae985436cbc942f0d11078041d121b where
  InitBlockIndex was writing to fTxIndex which had not yet been
  checked (because LoadChainTip hadn't yet initialized the
  chainActive, which would otherwise have resulted in
  InitBlockIndex being a NOP), allowing you to modify -txindex
  without reindex, potentially corrupting your chainstate!

* Rename InitBlockIndex to LoadGenesisBlock, which is now a more
  natural name for it. Also check mapBlockIndex instead of
  chainActive, fixing a bug where we'd write the genesis block out
  on every start.

TheBlueMatt committed:
More user-friendly error message if UTXO DB runs ahead of block DB

This gives LoadChainTip a return value - allowing it to indicate that
the UTXO DB ran ahead of the block DB. This just provides a nicer
error message instead of the previous mysterious
assert(!setBlockIndexCandidates.empty()) error.

This also calls ActivateBestChain in case we just loaded the genesis
block in LoadChainTip, avoiding relying on the ActivateBestChain
in ThreadImport before continuing init process.

TheBlueMatt commented:
Note that the two generic "Error initializing block database" strings really should be changed, but as we are already in string freeze for 0.15, they will have to stay as-is for now :(.

TheBlueMatt committed:
Call RewindBlockIndex even if we're about to run -reindex-chainstate

RewindBlockIndex works over both chainActive - disconnecting blocks
from the tip that need witness verification - and mapBlockIndex -
requiring redownload of blocks missing witness data.

It should never have been the case that the second half is skipped
if we're about to run -reindex-chainstate.

TheBlueMatt committed:
Order chainstate init more logically.

* Order chainstate init more logically - first all of the
  blocktree-related loading, then coinsdb, then
  pcoinsTip/chainActive. Only create objects as needed.

* More clearly document exactly what is and isn't called in
  -reindex and -reindex-chainstate both with comments noting
  calls as no-ops and by adding if guards.

* Move LoadGenesisBlock further down in init. This is a more logical
  location for it, as it is after all of the blockindex-related
  loading and checking, but before any of the UTXO-related loading
  and checking.

* Move all of the VerifyDB()-related stuff into a -reindex +
  -reindex-chainstate if guard. It couldn't do anything useful
  as chainActive.Tip() would be null at this point anyway.

TheBlueMatt commented:
More explicitly fixed the bug that I accidentally fixed by being more explicit - RewindBlockIndex MUST be called even if we start with -reindex-chainstate.

TheBlueMatt commented:
This needs an 0.15 tag as it fixes some bugs.

TheBlueMatt committed:
Fix segfault when shutting down before fully loading

This was introduced by 3192975f1d177aa9f0bbd823c6387cfbfa943610.
It can be triggered easily when canceling DB upgrade from
pre-per-utxo.

TheBlueMatt commented:
Rebased and added a fix for the issue @sipa pointed out at https://github.com/bitcoin/bitcoin/pull/10770#issuecomment-315558931

sipa commented:
@TheBlueMatt When starting with `-reindex`, and then restart before completing without `-reindex`, it seems to start over from scratch.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -3864,42 +3864,53 @@ void UnloadBlockIndex()
+    }
+
+    if (needs_init) {
+        // Note that LoadBlockIndexDB may have set fReindex if we shut down
```
OK, rewrote the comment.

sipa commented:
ACK 

@TheBlueMatt I retract my comment, everything seems to be working like before.

sipa commented:
I'd like to see this in 0.15 as it at least turns an assert into an error message, in case your chainstate is out of sync with the block chain.

TheBlueMatt commented:
This (or some subset of it) absolutely has to land for 15, it fixes several regressions currently on master, but its a bugfix, so doesn't need to land today (but isnt small, so should go sooner rather than later).

theuni commented:
I haven't finished reviewing, but this also fixes a crash when using an invalid -wallet argument. For example: ```./bitcoind -wallet='with spaces'```, which crashes for me on master.

morcos commented:
utACK c93db42

Strongly advocate for merging this in the near term.
Although I have not gone through testing all sequences, I think this logic is far clearer and more straight forward than the existing logic and there are known bugs in the existing startup sequence.
With all the other changes to databases recently (rewinding, replaying, reindexing), I think it makes sense to have as much testing as possible with this new code rather than waste our time potentially finding more bugs with the old logic.

promag commented:
In current master running `bitcoind -regtest '-wallet=foo%.dat'` gives:
```
./src/bitcoind -regtest '-wallet=foo%.dat'
Error: Invalid characters in -wallet filename
Segmentation fault (core dumped)
```
With this branch gives
```
Error: Invalid characters in -wallet filename
bitcoind: scheduler.cpp:200: void SingleThreadedSchedulerClient::EmptyQueue(): Assertion `!m_pscheduler->AreThreadsServicingQueue()' failed.
Aborted (core dumped)
```
Reporting this here because of c93db42.

TheBlueMatt commented:
Hmm, I went ahead and added the fix for @promag's race here, feel free to tell me it should be in a different PR. (@theuni note that it was a race, which may explain why you didnt see it).

TheBlueMatt commented:
@sipa hmm, re: restarting-mid-reindex: your original comment that it was broken was more correct than your later retraction. Fixed and cleaned that stuff up more with more comments :).

promag commented:
@TheBlueMatt if it's not related with your refactor then IMO a new PR would be best.

TheBlueMatt commented:
OK, removed the top two commits so this is just what it was that got ACKs. The two commits from the top are in #10919, which should likely also be taken for 15.

sipa commented:
reutACK c93db429f909e75fbeacd2419fb8434350e3b674

laanwj commented on src/validation.cpp:
```diff
@@ -3864,42 +3884,55 @@ void UnloadBlockIndex()
+        if (!ReceivedBlockTransactions(block, state, pindex, blockPos, chainparams.GetConsensus()))
+            return error("LoadBlockIndex(): genesis block not accepted");
+    } catch (const std::runtime_error& e) {
+        return error("LoadBlockIndex(): failed to initialize block database: %s", e.what());
```
This error is no longer correct, this is "failed to load genesis block" now I guess?

laanwj commented on src/validation.cpp:
```diff
@@ -3864,42 +3884,55 @@ void UnloadBlockIndex()
+        CDiskBlockPos blockPos;
+        CValidationState state;
+        if (!FindBlockPos(state, blockPos, nBlockSize+8, 0, block.GetBlockTime()))
+            return error("LoadBlockIndex(): FindBlockPos failed");
```
Would be better to use `error("%s: ...", __func__)` here as in other places, but in any case, the function name is wrong now.

laanwj commented:
utACK apart from error message nits

TheBlueMatt commented:
Fixed error printings in LoadGenesisBlock.

morcos commented:
re-utACK c0025d0

sdaftuar commented on src/validation.cpp:
```diff
@@ -3538,14 +3538,24 @@ bool static LoadBlockIndexDB(const CChainParams& chainparams)
     BlockMap::iterator it = mapBlockIndex.find(pcoinsTip->GetBestBlock());
     if (it == mapBlockIndex.end())
-        return;
+        return false;
```
style nit: perhaps add the curly braces here, while you're at it

sdaftuar commented on src/init.cpp:
```diff
@@ -1437,10 +1427,34 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
 
+                // At this point blocktree args are consistent with what's on disk.
+                // If we're not mid-reindex (based on disk + args), add a genesis block on disk.
+                // This is called again in ThreadImport in the reindex completes.
```
"in the reindex completes" -> "after the reindex completes"?

sdaftuar reviewed (approved):
utACK, left some nits.

Is it worth backporting the RewindBlockIndex bugfix to the 0.13 or 0.14 branch?  I could imagine someone upgrading from 0.13.0 to the latest 0.13 release after segwit activates.  Or to the latest 0.14 if the 0.15.0 release comes out after activation?

laanwj commented:
I'm going to merge this now - sorry @sdaftuar, we should fix your minor nits later, but we have to make some progress with merging for 0.15

sdaftuar commented:
@laanwj Agreed, I don't think those nits should have held this up.
